pkg/logs/sender/destination.go

```diff
@@ -32,26 +33,40 @@ func (e *FramingError) Error() string {
 
 // Destination is responsible for shipping logs to a remote server over TCP.
 type Destination struct {
-	prefixer    Prefixer
-	delimiter   Delimiter
-	connManager *ConnectionManager
-	conn        net.Conn
+	prefixer            Prefixer
+	delimiter           Delimiter
+	connManager         *ConnectionManager
+	destinationsContext *DestinationsContext
+	conn                net.Conn
 }
 
 // NewDestination returns a new destination.
-func NewDestination(endpoint config.Endpoint) *Destination {
+func NewDestination(endpoint config.Endpoint, destinationsContext *DestinationsContext) *Destination {
 	return &Destination{
-		prefixer:    NewAPIKeyPrefixer(endpoint.APIKey, endpoint.Logset),
-		delimiter:   NewDelimiter(endpoint.UseProto),
-		connManager: NewConnectionManager(endpoint),
+		prefixer:            NewAPIKeyPrefixer(endpoint.APIKey, endpoint.Logset),
+		delimiter:           NewDelimiter(endpoint.UseProto),
+		connManager:         NewConnectionManager(endpoint),
+		destinationsContext: destinationsContext,
 	}
 }
 
 // Send transforms a message into a frame and sends it to a remote server,
 // returns an error if the operation failed.
 func (d *Destination) Send(payload message.Message) error {
 	if d.conn == nil {
-		d.conn = d.connManager.NewConnection()
+		var ctx context.Context
+		var err error
+
+		// If we have a context, use it, this will allow early cancellation.
+		if d.destinationsContext != nil {
```

Comment:
I am not a huge fan of this, I think a sender should always have a `destinationsContext `

Reply:
So I looked again and .. The relationships between destinations senders and destinations context are quite complicated.

Currently, if there is a bug and destinationsContext isn't set properly, we just fail to cancel the destinations on shutdown. If I remove the check, we will crash. Can we keep the safer way for now, and re-evaluate this later ?

Reply:
👍